trillian/integration/copier.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/google/certificate-transparency-go/trillian/ctfe/configpb"
+	"github.com/google/certificate-transparency-go/x509"
+
+	ct "github.com/google/certificate-transparency-go"
+)
+
+// CopyChainGenerator creates certificate chains by copying suitable examples
+// from a source log.
+type CopyChainGenerator struct {
+	start, limit             time.Time
+	sourceRoots, targetRoots *ctfe.PEMCertPool
+	certs, precerts          chan []ct.ASN1Cert
+}
+
+// NewCopyChainGenerator builds a certificate chain generator that sources
+// chains from another source log, starting at startIndex (or a random index in
+// the current tree size if startIndex is -1).  This function starts background
+// goroutines that scan the log; cancelling the context will terminate these
+// goroutines (after that the [Pre]CertChain() entrypoints will permanently
+// fail).
+func NewCopyChainGenerator(ctx context.Context, client *client.LogClient, cfg *configpb.LogConfig, startIndex, bufSize int) (ChainGenerator, error) {
+	var start, limit time.Time
+	var err error
+	if cfg.NotAfterStart != nil {
+		start, err = ptypes.Timestamp(cfg.NotAfterStart)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse NotAfterStart: %v", err)
+		}
+	}
+	if cfg.NotAfterLimit != nil {
+		limit, err = ptypes.Timestamp(cfg.NotAfterLimit)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse NotAfterLimit: %v", err)
+		}
+	}
+
+	targetPool := ctfe.NewPEMCertPool()
+	for _, pemFile := range cfg.RootsPemFile {
+		if err := targetPool.AppendCertsFromPEMFile(pemFile); err != nil {
+			return nil, fmt.Errorf("failed to read trusted roots for target log: %v", err)
+		}
+	}
+	if glog.V(4) {
+		for _, cert := range targetPool.RawCertificates() {
+			glog.Infof("target root cert: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+		}
+	}
+
+	seenOverlap := false
+	srcRoots, err := client.GetAcceptedRoots(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read trusted roots for source log: %v", err)
+	}
+	sourcePool := ctfe.NewPEMCertPool()
+	for _, root := range srcRoots {
+		cert, err := x509.ParseCertificate(root.Data)
+		if x509.IsFatal(err) {
+			glog.Warningf("Failed to parse root certificate from source log: %v", err)
+			continue
+		}
+		glog.V(4).Infof("source log root cert: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+		sourcePool.AddCert(cert)
+		if targetPool.Included(cert) {
+			glog.V(3).Infof("source log root cert is accepted by target: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+			seenOverlap = true
+		}
+	}
+	if !seenOverlap {
+		return nil, fmt.Errorf("failed to find any overlap in accepted roots for target %s", cfg.Prefix)
+	}
+
+	if startIndex == -1 {
+		// Pick a random start point in the source log's tree.
+		sth, err := client.GetSTH(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get STH for source log: %v", err)
+		}
+		startIndex = rand.Intn(int(sth.TreeSize))
```

Comment:
Don' think that that int conversion is safe here. It could be 32 bits and we have test trees bigger than that. Also can it go negative?

Reply:
Moved to `int64`; it can be negative on entry to indicate a random start point, but should be >=0 by the time it gets passed to `scanner.NewFetcher` (but which expects it as an `int64` not `uint64`)